internal/mongo/watch_producer.go

```diff
@@ -31,49 +32,95 @@ func (w *WatchProducer) GetProducer(o ...WatchOption) ChangeEventProducer {
 			pipeline = append(customElements, pipeline...)
 		}
 
-		opts := &options.ChangeStreamOptions{
-			BatchSize:            &config.batchSize,
-			MaxAwaitTime:         &config.maxAwaitTime,
-			StartAtOperationTime: config.startAtOperationTime,
-		}
-		if config.fullDocumentEnabled {
-			opts.SetFullDocument(options.UpdateLookup)
-		}
-		if len(config.resumeAfter) != 0 {
-			opts.SetResumeAfter(config.resumeAfter)
-		}
-
-		cursor, err := w.collection.Watch(ctx, pipeline, opts)
+		cursor, err := w.watch(ctx, pipeline, config, nil)
 		if err != nil {
-			w.logger.Error("Mongo client: An error has occured while watching collection", logger.String("collection", w.collection.Name()), logger.Error("error", err))
+			w.logger.Error("Mongo client: An error has occured while trying to watch collection", logger.String("collection", w.collection.Name()), logger.Error("error", err))
 			return nil, err
 		}
 
 		var events = make(chan *ChangeEvent)
 
 		go func() {
-			defer cursor.Close(ctx)
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
 			defer close(events)
-
 			for {
-				w.sendEvents(ctx, cursor, events)
+				resumeToken, err := w.sendEvents(ctx, cursor, events)
+				w.logger.Error("Mongo client : An error has occured while watching collection", logger.String("collection", w.collection.Name()), logger.Error("error", err))
+
+				cursor.Close(ctx)
+
+				w.logger.Info("Mongo client : Retry to watch collection", logger.String("collection", w.collection.Name()), logger.Any("token", resumeToken))
+
+				cursor, err = w.watch(ctx, pipeline, config, resumeToken)
+				if err != nil {
+					w.logger.Error("Mongo client : An error has occured while retrying to watch collection", logger.String("collection", w.collection.Name()), logger.Error("error", err))
+					break
+				}
 			}
 		}()
 
 		return events, nil
 	}
 }
 
-func (w *WatchProducer) sendEvents(ctx context.Context, cursor DriverCursor, events chan *ChangeEvent) {
+func (w *WatchProducer) watch(ctx context.Context, pipeline bson.A, config *WatchConfig, resumeToken bson.Raw) (cursor StreamCursor, err error) {
+	// choose the resume token to use
+	var resumeAfter interface{}
+	if resumeToken != nil {
+		resumeAfter = resumeToken
+	} else if len(config.resumeAfter) > 0 {
+		resumeAfter = config.resumeAfter
+	}
+	// retries loop
+	attempt := int32(0)
+	for {
+		opts := &options.ChangeStreamOptions{
```

Comment:
we can use here a backoff retry lib like : https://github.com/cenkalti/backoff/blob/v4/example_test.go#L23

Reply:
I'm not sure we need to add a dependency just for that.

Reply:
It is not just a dependency it is a behaviour for the recovery strategy. 
How i see the thing is that it should keep retrying to connect until it is a success, and of course you cannot do an infinite loop with 250 ms of wait.
That is why a lib like that is interesting, it keeps trying quickly for a short period of time and after that the interval is bigger and finally reconnects when the db is up again without restarting the pod.

Reply:
I see. Could you please add this lib in another PR ?